### DISCUSSION TITLE
JC4827W543R OK in Arduino_GFX/QSPI, how to port to Lovyan?

### DISCUSSION THREAD

**Discussion by Westcott1** [lovyan03/LovyanGFX#604] [Device Configurations]
I've just got a JC4827W543C 4.3" 480x272 "cheap black display" (CBD).
[https://github.com/profi-max/JC4827W543_4.3inch_ESP32S3_board](https://github.com/profi-max/JC4827W543_4.3inch_ESP32S3_board)
It's working OK with Arduino_GFX, config below.
Can this translate to LovyanGFX?

```
#include <Arduino_GFX_Library.h>

// display information
#define TFT_WIDTH 480
#define TFT_HEIGHT 272
#define TFT_ORIENTATION 2

// setup Arduino_GFX
#define GFX_BL 1

static Arduino_ESP32QSPI bus{
  45 /* cs */,
  47 /* sck */,
  21 /* d0 */,
  48 /* d1 */,
  40 /* d2 */,
  39 /* d3 */
};
static Arduino_NV3041A gfx{
  &bus,
  GFX_NOT_DEFINED /* RST */,
  TFT_ORIENTATION,
  true /* IPS */
};
```

**Comment by tobozo** [comment]
hi,

LovyanGFX does not support QSPI for ESP32 **yet**, you can still try this [PR](https://github.com/lovyan03/LovyanGFX/pull/388) if you feel experimental though :-)

Using the QSPI implementation from that PR, your config file should look like [this](https://github.com/Forairaaaaa/Monica/blob/main/Firmware/SourceCode/main/hal/disp/hal_disp.hpp)

[edit1] updated the link to PR
[edit2] added link to QSPI config example

**Reply by Westcott1** [reply]
Hi Tobozo,
Thanks for the reply and example!
I'll give it a go real soon.
UPDATE
I got the complete Monica source for lgfx but was unable to compile it.
It looks like a fork from an earlier version of lgfx.

**Reply by tobozo** [reply]
the PR was merged on a [separate branch](https://github.com/lovyan03/LovyanGFX/tree/QSPI), it compiles but remains untested

**Reply by Westcott1** [reply]
Thanks!
I've downloaded and installed this branch.
My code now compiles - after I added `#include "v1/panel/Panel_SH8601Z.hpp"` to v1_init.hpp.

However, it hangs in `gfx.init()` with -
E (11173) spi_master: spi_master_init_driver(236): invalid core id

I'm using "ESP32 Dev Module" from boards esp32 v3.0.4
Here's my config, touch and backlight work OK (sorry I don't know what's wrong with the Code formatting) -

```cpp

// This is for a Guition JC4827W543C
// 4.3" 480 x 270
#pragma once

#define LGFX_USE_V1

#include <LovyanGFX.hpp>

#define TFT_HOR_RES 480
#define TFT_VER_RES 272

#define TFT_SCLK 47
#define TFT_CS 45
#define TFT_RST 4

#define TFT_BL 1

#define TOUCH_IRQ -1
#define TOUCH_SCL 4
#define TOUCH_SDA 8

class LGFX : public lgfx::LGFX_Device {
  lgfx::Panel_SH8601Z _panel_instance;
  lgfx::Bus_QSPI _bus_instance;
  lgfx::Touch_GT911 _touch_instance;
  lgfx::Light_PWM _light_instance;

public:
  LGFX(void) {
    {
      auto cfg = _bus_instance.config();

      cfg.spi_host = SPI3_HOST;
      cfg.spi_mode = 1;
      cfg.freq_write = 40000000;
      cfg.freq_read = 16000000;
      cfg.spi_3wire = true;
      cfg.use_lock = true;
      cfg.dma_channel = SPI_DMA_CH_AUTO;

      cfg.pin_sclk = TFT_SCLK;  // SPI SCK pin
      cfg.pin_io0 = 21;
      cfg.pin_io1 = 48;
      cfg.pin_io2 = 40;
      cfg.pin_io3 = 39;

      _bus_instance.config(cfg);
      _panel_instance.setBus(&_bus_instance);
    }

    {
      auto cfg = _panel_instance.config();

      cfg.pin_cs = TFT_CS;    // Chip select pin
      cfg.pin_rst = TFT_RST;  // Reset pin
      cfg.pin_busy = -1;      // Busy pin (not used)

      cfg.panel_width = TFT_HOR_RES;
      cfg.panel_height = TFT_VER_RES;
      cfg.offset_x = 0;
      cfg.offset_y = 0;
      cfg.offset_rotation = 0;
      cfg.dummy_read_pixel = 8;
      cfg.dummy_read_bits = 1;
      cfg.readable = true;
      cfg.invert = true;
      cfg.rgb_order = true;
      cfg.dlen_16bit = false;
      cfg.bus_shared = true;
      cfg.memory_width = TFT_HOR_RES;
      cfg.memory_height = TFT_VER_RES;

      _panel_instance.config(cfg);
    }

    {  // Configure settings for touch screen control.  (delete if not necessary)
      auto cfg = _touch_instance.config();

      cfg.x_min = 0;
      cfg.x_max = TFT_HOR_RES - 1;
      cfg.y_min = 0;
      cfg.y_max = TFT_VER_RES - 1;
      cfg.pin_int = -1;        // Pin number to which INT is connected (-1 = not connected)
      cfg.bus_shared = false;  // Set true when using a common bus with the screen
      // cfg.offset_rotation = 2;  // Adjustment when the orientation does not match Set with a value from 0 to 7
      cfg.offset_rotation = 6;  // 6 fits but double lines. why?

      // For SPI connection
      cfg.spi_host = SPI2_HOST;  // Select SPI to use (HSPI_HOST or VSPI_HOST)
      cfg.freq = 2500000;        // SPI Clock frequency 1000000 -> 2500000
      cfg.pin_sda = TOUCH_SDA;   // SCLK
      cfg.pin_scl = TOUCH_SCL;  // CS

      _touch_instance.config(cfg);
      _panel_instance.setTouch(&_touch_instance);
    }

    {
      auto cfg = _light_instance.config();

      cfg.pin_bl = TFT_BL;  // Backlight pin
      cfg.invert = false;

      _light_instance.config(cfg);
      _panel_instance.setLight(&_light_instance);
    }

    setPanel(&_panel_instance);
  }
  
  inline bool init(void) {
    // PEN pin 
    gpio_reset_pin(GPIO_NUM_4);
    gpio_set_direction(GPIO_NUM_4, GPIO_MODE_OUTPUT);
    gpio_set_pull_mode(GPIO_NUM_4, GPIO_PULLUP_PULLDOWN);
    gpio_set_level(GPIO_NUM_4, 1);
    // vTaskDelay(pdMS_TO_TICKS(10));

    // TE pin
    gpio_reset_pin(GPIO_NUM_2);
    gpio_set_direction(GPIO_NUM_2, GPIO_MODE_INPUT);
    // vTaskDelay(pdMS_TO_TICKS(10));

    // Lgfx
    return lgfx::LGFX_Device::init();
  }
};
```

**Comment by tobozo** [comment]
your QSPI display is a [NV3041A](https://github.com/profi-max/JC4827W543_4.3inch_ESP32S3_board/blob/main/Examples/Demo_Arduino/3_3-1_TFT_PDQgraphicstest/PDQgraphicstest/Arduino_GFX_dev_device.h#L109C29-L109C37), not a SH8601Z.

the reason `Panel_SH8601Z` has been imported with the PR is because it is the only QSPI panel example, that class won't work with a NV3041A display.

what is needed now is the creation of a `Panel_NV3041A` class.

solving the [E (11173) spi_master: spi_master_init_driver(236): invalid core id](https://github.com/espressif/esp-idf/blob/master/components/esp_driver_spi/src/gpspi/spi_master.c#L297) bus error (display task is running on the wrong core): either change the core affinity setting, or attach the display task to the other core, also make sure all tasks using SPI are on the same core.

I have produced the skeleton of a [`Panel_NV3041A` class](https://github.com/lovyan03/LovyanGFX/commit/85309172b6c9cce04a151841a05ae73f9b1330ef), it compiles but remains untested.

```cpp
#include <LovyanGFX.hpp>
#include <lgfx_user/LGFX_JC4827W543_4.3inch_ESP32S3.hpp>

LGFX tft;

void setup()
{
  tft.init();
}

void loop()
{
  tft.clear();
  tft.println("Hello World!");
  delay(1000);
}
```

**Reply by Westcott1** [reply]
Wow Tobozo!
Thanks for all of that!

As for `E (11173) spi_master: spi_master_init_driver(236): invalid core id.`
This only occurs if I assign any non-negative number to **`cfg.pin_io2`** in the QSPI _bus_instance.
-1 is OK.
io0, i01, and io3 are also OK for their correct values..

**Comment by Zurrmaxe** [comment]
Hello Friends
I have the same Board and want to use it with lovyanGFX.(but without Touch).
Did anyone have a working sketch like "Hello World" ?
I tried the modifications from this side, but nothing worked.
Alf

**Reply by Westcott1** [reply]
Hi Zurrmaxe,
This board won't work with LovyanGFX until Expressive ESP32 3.1.0 for Arduino.
Mine works OK with ArduinoGFX.

`#include <Arduino_GFX_Library.h>

// display information
#define TFT_WIDTH 480
#define TFT_HEIGHT 272
#define TFT_ORIENTATION 2

// LCD backlight PWM 
#define LCD_BL 1           // lcd BL pin

Arduino_DataBus *bus = new Arduino_ESP32QSPI{
  45 /* cs */,
  47 /* sck */,
  21 /* d0 */,
  48 /* d1 */,
  40 /* d2 */,
  39 /* d3 */
};

Arduino_GFX *gfx = new Arduino_NV3041A {
  bus,
  GFX_NOT_DEFINED, // RST
  TFT_ORIENTATION,
  true // IPS
};
`

**Comment by tobozo** [comment]
closing this as [implemented](https://github.com/lovyan03/LovyanGFX/blob/develop/src/lgfx_user/LGFX_JC4827W543_4.3inch_ESP32S3.hpp)

@Westcott1 bugs in panel support (e.g. rgb order) will be handled in your [other issue](https://github.com/lovyan03/LovyanGFX/issues/609)
